File: frontend/src/api/notebook.ts
import client from './client'

export interface Notebook {
  id: string
  user_id: string
  title: string
  description: string
  created_at: string
  updated_at: string
  source_count: number
  share_enabled?: boolean
}

export interface SubscribedNotebookItem {
  notebook: Notebook
  read_available: boolean
  share_token: string | null
}

interface NotebookSubscriptionsResponse {
  items: SubscribedNotebookItem[]
  total: number
}

interface NotebookListResponse {
  notebooks: Notebook[]
  total: number
}

export const notebookApi = {
  list: async (): Promise<NotebookListResponse> => {
    const res = await client.get('/notebooks')
    return res.data
  },

  get: async (id: string): Promise<Notebook> => {
    const res = await client.get(`/notebooks/${id}`)
    return res.data
  },

  create: async (data: { title: string; description?: string }): Promise<Notebook> => {
    const res = await client.post('/notebooks', data)
    return res.data
  },

  update: async (id: string, data: { title?: string; description?: string }): Promise<Notebook> => {
    const res = await client.put(`/notebooks/${id}`, data)
    return res.data
  },

  remove: async (id: string): Promise<void> => {
    await client.delete(`/notebooks/${id}`)
  },

  enableShare: async (
    id: string,
    body: { regenerate?: boolean } = {},
  ): Promise<{ share_token: string }> => {
    const res = await client.post(`/notebooks/${id}/share`, body)
    return res.data
  },

  disableShare: async (id: string): Promise<void> => {
    await client.delete(`/notebooks/${id}/share`)
  },

  listPublished: async (): Promise<NotebookListResponse> => {
    const res = await client.get<NotebookListResponse>('/notebooks/published')
    return res.data
  },

  listSubscriptions: async (): Promise<NotebookSubscriptionsResponse> => {
    const res = await client.get<NotebookSubscriptionsResponse>(
      '/notebooks/subscriptions',
    )
    return res.data
  },

  publishToDiscover: async (
    id: string,
    body: { category?: string; cover_url?: string } = {},
  ): Promise<void> => {
    await client.post(`/notebooks/${id}/discover/publish`, body)
  },

  unpublishFromDiscover: async (id: string): Promise<void> => {
    await client.delete(`/notebooks/${id}/discover/publish`)
  },
}

File: backend/app/schemas/notebook.py
"""Pydantic schemas for notebooks."""

from datetime import datetime

from pydantic import BaseModel


class NotebookCreate(BaseModel):
    """Schema for creating a notebook."""

    title: str
    description: str = ""


class NotebookUpdate(BaseModel):
    """Schema for updating a notebook."""

    title: str | None = None
    description: str | None = None


class NotebookResponse(BaseModel):
    """Schema for notebook response."""

    id: str
    user_id: str
    title: str
    description: str
    created_at: datetime
    updated_at: datetime
    source_count: int = 0
    share_enabled: bool = False

    model_config = {"from_attributes": True}


class NotebookShareBody(BaseModel):
    """Enable or rotate public share link."""

    regenerate: bool = False


class NotebookShareLinkResponse(BaseModel):
    """Returned when creating or rotating a share token."""

    share_token: str


class SharedNotebookView(BaseModel):
    """Notebook metadata exposed to anonymous share viewers."""

    id: str
    title: str
    description: str
    created_at: datetime
    updated_at: datetime
    source_count: int = 0


class NotebookListResponse(BaseModel):
    """Schema for notebook list response."""

    notebooks: list[NotebookResponse]
    total: int


class NotebookSubscriptionItem(BaseModel):
    """Subscribed notebook row for the home subscriptions tab."""

    notebook: NotebookResponse
    read_available: bool
    share_token: str | None = None


class NotebookSubscriptionsListResponse(BaseModel):
    """List of notebooks the current user subscribed to."""

    items: list[NotebookSubscriptionItem]
    total: int
